Index: proxpatches/rootfs/usr/share/perl5/PVE/VZDump/LXC.pm
===================================================================
--- proxpatches.orig/rootfs/usr/share/perl5/PVE/VZDump/LXC.pm
+++ proxpatches/rootfs/usr/share/perl5/PVE/VZDump/LXC.pm
@@ -327,7 +327,7 @@ sub archive {
     my $tmpdir = $task->{tmpdir};
 
     my $userns_cmd = $task->{userns_cmd};
-    my $tar = [@$userns_cmd, 'tar', 'cpf', '-', '--totals',
+    my $tar = [@$userns_cmd, 'tar', 'cpf', '-', '--totals', '--exclude-from=/etc/vzdump-excludes',
                @$PVE::LXC::COMMON_TAR_FLAGS,
                '--one-file-system', '--warning=no-file-ignored'];
 
